flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/table/DynamicKafkaSerializationSchema.java

```diff
@@ -166,7 +168,7 @@ public void setPartitions(int[] partitions) {
 
     @Override
     public String getTargetTopic(RowData element) {
-        return topic == null
+        return TOPIC_UNSPECIFIED.equals(topic)
                 ? readMetadata(element, KafkaDynamicSink.WritableMetadata.TOPIC)
```

Comment:
I just saw that `readMetadata` can return `null` which it shouldn't (and it's not annotated!). 
What is the expected target topic then? I guess `TOPIC[0]`?
But what happens if the user didn't set `TOPIC`? Can you check the schema if the column is nullable and fail if `TOPIC` is empty/unset?

Reply:
This comment is still not addressed. When neither option nor metadata column are defined, `readMetadata` will return null and cause Kafka to throw an exception. We can be smarter here and fail already during planning. In `KafkaDynamicSink#getSinkRuntimeProvider` we have all information to create a helpful exception if the metadata column is missing.